Poor-man e2e tests

inercia opened the pull request:
It would be good to have some simple e2e tests for doing things like creating a cluster, doing some tests and removing the cluster.

Depends on #169

iwilltry42 commented on Makefile:
```diff
@@ -62,6 +62,9 @@ LINT_DIRS := $(DIRS) $(foreach dir,$(REC_DIRS),$(dir)/...)
 all: clean fmt check build
 
 build:
+	$(GO) build -i $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' -o '$(BINDIR)/$(BINARIES)'
```
Is there a specific reason to use the CGO_ENABLED build as the default?

inercia replied:
```bash
$ make build-nocgo 
CGO_ENABLED=0 go build -i  -tags '' -ldflags '-w -s -X github.com/rancher/k3d/version.Version=v1.4.0-11-g407b153 -X github.com/rancher/k3d/version.K3sVersion=v1.17.0-k3s.1' -o '/home/alvaro/Development/k3d/bin/k3d'
go build net: open /usr/lib/go-1.13/pkg/linux_amd64/net.a: permission denied
Makefile:68: recipe for target 'build-nocgo' failed
make: *** [build-nocgo] Error 1
```

iwilltry42 replied:
This seems to be related to the `-i` flag.
After reading through [this issue](https://github.com/golang/go/issues/24674), I think we should rather consider removing the `-i` than switching between CGO and non-CGO (also see https://github.com/golang/go/issues/24674#issuecomment-379920545).

iwilltry42 replied:
Actually, according to comments on [this isse](https://github.com/golang/go/issues/25138), removing the `-i` flag in projects which use Go >= 1.10 is the better option. 
According to [the last comment there](https://github.com/golang/go/issues/25138#issuecomment-490621200), they're even going to disable installing stdlib dependencies if `-i` is set in a future release of Go.

inercia replied:
Cool. Fixed.

iwilltry42 commented on Makefile:
```diff
@@ -110,3 +120,17 @@ else
+
+ci-setup:
+	@echo "Installing Go..."
+	curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b ${GOPATH}/bin v1.20.0
```
We have variables for the `gox` and `golangci-lint` versions that we can re-use here :+1:

iwilltry42 commented on tests/01-basic.sh:
```diff
@@ -0,0 +1,31 @@
+
+function finish {
+  info "Cleaning up any leftovers..."
+  $EXE delete --all > /dev/null 2>&1 || /bin/true
```
This can be dangerous if someone is running this on his/her local machine (not inside dind).

inercia replied:
I see... Then I will remove it. At some moment it would be good to have some `--filter` argument...

iwilltry42 commented on Makefile:
```diff
@@ -110,3 +120,17 @@ else
 endif
+
+ci-setup:
+	@echo "Installing Go..."
```
You're not really installing go here, right? :grin:

inercia replied:
Right. I'll say "Go tools"

iwilltry42 commented:
It would be cool to have all tests running in a docker container :thinking:

inercia commented:
> It would be cool to have all tests running in a docker container

I have tried to do that, but it seems it cannot find the Docker daemon...

iwilltry42 commented:
> > It would be cool to have all tests running in a docker container
> 
> I have tried to do that, but it seems it cannot find the Docker daemon...

How did you try it? It may even be a good idea to slip all the building etc. into containers, so that we definitely end up with the same builds on every machine :thinking: 
But for now, testing in containers would be great already :+1:

inercia commented:
> > > It would be cool to have all tests running in a docker container
> > 
> > 
> > I have tried to do that, but it seems it cannot find the Docker daemon...
> 
> How did you try it? It may even be a good idea to slip all the building etc. into containers, so that we definitely end up with the same builds on every machine 
> But for now, testing in containers would be great already

I have fixed the first problem, but now we cannot run the e2e tests inside because the `kubeconfig` generated by kl3s/k3d points to `localhost`, but when the kubectl in the DinD container tries to access `localhost` it fails... We would need to get the host IP and use it in all the `k3d create --api-port <HOST_IP>`...

 I have added [a `e2e-dind` target](https://github.com/rancher/k3d/pull/168/files#diff-b67911656ef5d18c4ae36cb6741b7965R91-R100).

iwilltry42 commented:
I updated the PR a bit to get the tests working in dind. Now it first creates the k3d-dind container, waits for a few seconds (to give docker time to start) and then uses `docker exec` to run the tests inside the container. This is now working on my local machine and in Travis CI ([see e.g. this build]()https://travis-ci.com/rancher/k3d/builds/145318830).

iwilltry42 commented:
Awesome, thanks a lot for this :)
